miner/miner.go

```diff
@@ -106,11 +106,12 @@ func (miner *Miner) update() {
 					log.Info("Mining aborted due to sync")
 				}
 			case downloader.DoneEvent, downloader.FailedEvent:
-				canStart = true
 				if shouldStart {
 					miner.SetEtherbase(miner.coinbase)
 					miner.worker.start()
 				}
+				// stop immediately and ignore all further pending events
+				return
```

Comment:
I think this fix is a bit too crude. 

- If you want to start mining after having synced, you cannot
- If you start mining before sync, but your initial sync fails/aborts, then it will never start mining even if it later syncs successfully
- If you want to stop mining, sorry, can't do that

And it will never close the worker channels if you try to exit geth.

Reply:
Oh, I see.. So none of these cases are handled by the tests? I'll see what I can do this morning.

Reply:
Basically, this loop is the governor of mining. It decides whether mining happens or not, so if you want to change the behavior, that's one thing, but exiting it is not a good idea.

Reply:
Understood. I was confused because the `update` method's comment reading:

```go
// update keeps track of the downloader events. Please be aware that this is a one shot type of update loop.
// It's entered once and as soon as `Done` or `Failed` has been broadcasted the events are unregistered and
// the loop is exited. This to prevent a major security vuln where external parties can DOS you with blocks
// and halt your mining operation for as long as the DOS continues.
```

 was unchanged to reflect the new expected behavior introduced at https://github.com/ethereum/go-ethereum/pull/21536, the notes of which also lack any detail or comment on the revised use of the method. Shoulda read the code, ignored the words ;)